### PR TITLE
Refactored uint64_t SearchSignature to std::vector<AttributeConstraint>

### PR DESCRIPTION
**Pull request by SamArch27**
Previously we used the uint64_t data type to represent a SearchSignature. If a search where to have arity > 64 then an assert would be fired since we wouldn't have enough bits to support this situation.

To rectify this, we introduce a vector of AttributeConstraints { None, Equal, Inequality } which can have an arbitrary arity. This refactor involved the following:

1. Change all usages of SearchSignature to conform with the new SearchSignature interface.
2. Add tests to ensure that after this refactor, a minimum index selection can still be performed correctly when the SearchSignature has >64 bits.

Note: The old implementation relied on setting the MSB and flagging Signatures with the 32nd bit in order to identify whether it was on side A or side B of the bipartition. After my refactor, we have removed any dependency on having 64 bits and instead we use explicit maps to identify which bipartition each SearchSignature belongs to.

The mapping concept is like: Searches = { s1, s2, s3, s4 ..., sN }. When mapping to nodes we have:    s1 ->  Node 1 in  partition A
s1 ->  Node 2 in partition B
s2 -> Node 3 in partition A
s2 -> Node 4 in partition B
...
This mapping scheme continues for all of the nodes.

### CODE REVIEW

**Comment by b-scholz** [comment]
This is related to issue #995

**Comment by b-scholz** [comment]
Jenkins test please.

**Comment by b-scholz** [comment]
Jenkins, test this please

File: src/Synthesiser.cpp [hunk 1]
```diff
@@ -130,20 +130,21 @@ void Synthesiser::generateRelationTypeStruct(
...
+            if (!first) {
                 tmp << ",";
+            } else {
+                first = false;
             }
+            tmp << i;
         }
-        key >>= 1;
-        i++;
     }
```
Comment (ohamel-softwaresecure) [inline src/Synthesiser.cpp:148]: Simplify: `tmp << join(key); if (!key.empty()) tmp << ",";`
(Matching proposed change. IDK if you really need/want that tailing `,` if non-empty.)

File: src/RamIndexAnalysis.h [hunk 2]
```diff
@@ -34,16 +34,160 @@
...
+        return s >>= n;
+    }
+
+    inline size_t arity() const {
+        return bits.size();
+    }
+    friend std::ostream& operator<<(std::ostream& out, const SearchSignature& signature);
+
+private:
+    std::vector<bool> bits;
```
Comment (ohamel-softwaresecure) [inline src/RamIndexAnalysis.h:180]: `std::vector<bool>` is no longer required to be a template specialization in C++17, it is impl defined. If you want an actual bitset, you'll have to hand roll this yourself. Trivial enough, use a `uint32_t` as an element type and you'll likely get the small-vector-optimization to kick in (if impl in your STL).

File: src/RamIndexAnalysis.h [hunk 3]
```diff
@@ -34,16 +34,160 @@
...
+ * no value exists (i.e. attribute is unbounded) in the search. */
+class SearchSignature {
+public:
+    explicit SearchSignature(size_t arity, uint64_t mask) : bits(arity, false) {
+        size_t len = bits.size();
+        for (size_t i = 0; i < len; ++i) {
+            bits[i] = (mask % 2);
+            mask /= 2;
+        }
+    }
```
Comment (ohamel-softwaresecure) [inline src/RamIndexAnalysis.h:55]: This is effectively `boost::dynamic_bitset` (or something near to it). Could you rename it to `dynamic_bitset`, move it to its own header, and add an alias `using SearchSignature = dynamic_bitset`? It could be handy on its own.

Adding basic iterators support would make this compatible with the many existing helpers (e.g. `join`).

File: src/RamIndexAnalysis.h [hunk 4]
```diff
@@ -34,16 +34,160 @@
...
+    inline bool operator>(const SearchSignature& other) const {
+        return other < *this;
+    }
+    inline bool operator==(const SearchSignature& other) const {
+        assert(bits.size() == other.bits.size());
+        return bits == other.bits;
+    }
+    inline bool operator!=(const SearchSignature& other) const {
+        return !(other == *this);
+    }
```
Comment (ohamel-softwaresecure) [inline src/RamIndexAnalysis.h:89]: Potentially confusing behaviour: What happens in the case where `this->bits` and `rhs.bits` have different lengths? e.g. `10` vs. `100`? The details of the backing store should be opaque to the user, does it canonicalize after every operation? This is a bit-set and should behave as if it were a `set<int>` IMO.

File: src/RamIndexAnalysis.h [hunk 5]
```diff
@@ -34,16 +34,160 @@
...
+public:
+    explicit SearchSignature(size_t arity, uint64_t mask) : bits(arity, false) {
+        size_t len = bits.size();
+        for (size_t i = 0; i < len; ++i) {
+            bits[i] = (mask % 2);
+            mask /= 2;
+        }
+    }
+    // array subscript operator
+    inline bool operator[](std::size_t pos) const {
```
Comment (ohamel-softwaresecure) [inline src/RamIndexAnalysis.h:57]: Remove all these `inline`s. This qualifier has no effect on template function definitions nor member function definitions within an aggregate definition.

File: src/RamIndexAnalysis.h [hunk 6]
```diff
@@ -34,16 +34,160 @@
...
 namespace souffle {
 
 class RamTranslationUnit;
 
+/** search signature of a RAM operation; each bit represents an attribute of a relation.
+ * A one represents that the attribute has an assigned value; a zero represents that
+ * no value exists (i.e. attribute is unbounded) in the search. */
+class SearchSignature {
+public:
+    explicit SearchSignature(size_t arity, uint64_t mask) : bits(arity, false) {
```
Comment (ohamel-softwaresecure) [inline src/RamIndexAnalysis.h:49]: Add `default`ed copy/move ctors/assigners?

File: src/Synthesiser.cpp [hunk 7]
```diff
@@ -850,9 +851,9 @@ void Synthesiser::emitCode(std::ostream& out, const RamStatement& stmt) {
 
             // get range to aggregate
             auto keys = isa->getSearchSignature(&aggregate);
-
+            SearchSignature zero(arity, 0);
             // special case: counting number elements over an unrestricted predicate
-            if (aggregate.getFunction() == AggregateOp::COUNT && keys == 0 &&
+            if (aggregate.getFunction() == AggregateOp::COUNT && keys == zero &&
```
Comment (ohamel-softwaresecure) [inline src/Synthesiser.cpp:856]: Simplify to `keys.empty()`.

File: src/RamIndexAnalysis.h (hunk 6 above)
Comment (ohamel-softwaresecure) [reply src/RamIndexAnalysis.h:49]: Remove `arity` argument, the size of the storage should be an internal detail.

File: src/test/max_matching_test.cpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2013, 2015, Oracle and/or its affiliates. All rights reserved
```
Comment (mmcgr) [inline src/test/max_matching_test.cpp:3]: Unless you're just renaming an existing test, this should be 2020, The Souffle Developers

**Comment by mmcgr** [comment]
Jenkins, test this please

File: src/Synthesiser.cpp (hunk 1 above)
Comment (SamArch27) [reply src/Synthesiser.cpp]: This wasn't called anywhere and I just use the overloaded ostream operator to print the SearchSignature in these contexts (such as lowerUpperRange_x etc.) therefore I just deleted it. Good point for future reference that format() is the more efficient idiom.

File: src/RamIndexAnalysis.h (hunk 3 above)
Comment (SamArch27) [reply src/RamIndexAnalysis.h]: I've implemented this. We have moved away from the bitset/dynamic bitset like format to a much nicer vector<AttributeConstraint> where AttributeConstraint is an enum representing the states that Bernhard described above.

File: src/RamIndexAnalysis.h (hunk 5 above)
Comment (SamArch27) [reply src/RamIndexAnalysis.h]: I have everything in a header so I used inline to get around ODR. Should I move it to a proper .cpp source file or just keep the inline?

File: src/Synthesiser.cpp (hunk 7 above)
Comment (SamArch27) [reply src/Synthesiser.cpp]: I did this. Great suggestion.

File: src/RamIndexAnalysis.h (hunk 6 above)
Comment (SamArch27) [reply src/RamIndexAnalysis.h]: I feel like the arity is intrinsic to the nature of the SearchSignature representing arity constraints (a constraint for each attribute). Given that we now have a vector<AttributeConstraint> perhaps the rarity in the constructor makes sense now?

File: src/test/max_matching_test.cpp (hunk 8 above)
Comment (SamArch27) [reply src/test/max_matching_test.cpp]: Fixed. Great catch.

**Comment by SamArch27** [comment]
Regarding the concern about having overloaded ==, <, > operators etc. I believe we need these to support SearchSignature being in STL containers such as std::set.

To update those watching this PR: I've refactored this once more to a more appropriate design which instead relies on an enum AttributeConstraint which can be either None, Equal or Inequal. The SearchSignature instead of relying on bits, now relies on a std::vector<AttributeConstraint> for its underlying representation.

I'd appreciate any and all feedback!

File: src/SynthesiserRelation.cpp [hunk 9]
```diff
@@ -1163,7 +1163,14 @@ void SynthesiserEqrelRelation::generateTypeStruct(std::ostream& out) {
...
     size_t arity = 2;
     for (int i = 1; i < 4; i++) {
-        SearchSignature s(arity, i);
+        SearchSignature s(arity);
+        if (i & 1) {
+            s.set(0, AttributeConstraint::Equal);
+        }
+        if (i & 2) {
+            s.set(1, AttributeConstraint::Equal);
+        }
```
Comment (ohamel-softwaresecure) [inline src/SynthesiserRelation.cpp:1172]: I had to stare at this for 15+ seconds. Maybe too clever?

File: src/test/matching_test.cpp [hunk 10]
```diff
@@ -70,26 +74,14 @@ TEST(Matching, StaticTest_2) {
...
-    order.addSearch(SearchSignature(arity, 32));
-    nodes.insert(SearchSignature(arity, 32));
-    order.addSearch(SearchSignature(arity, 23));
-    nodes.insert(SearchSignature(arity, 23));
-    order.addSearch(SearchSignature(arity, 11));
-    nodes.insert(SearchSignature(arity, 11));
-    order.addSearch(SearchSignature(arity, 7));
-    nodes.insert(SearchSignature(arity, 7));
+
+    std::array<uint64_t, 10> patterns = {7, 11, 23, 32, 33, 39, 49, 53, 104, 121};
```
Comment (ohamel-softwaresecure) [inline src/test/matching_test.cpp:78]: you can shorten this (and other instances of this) to `uint64_t patterns[] { .... };`

File: src/test/matching_test.cpp [hunk 11]
```diff
@@ -40,24 +40,28 @@ class TestAutoIndex : public MinIndexSelection {
...
-    nodes.insert(SearchSignature(arity, 5));
-    order.addSearch(SearchSignature(arity, 3));
-    nodes.insert(SearchSignature(arity, 3));
-    order.addSearch(SearchSignature(arity, 1));
-    nodes.insert(SearchSignature(arity, 1));
+
+    std::array<uint64_t, 7> patterns = {1, 3, 5, 7, 15, 23, 31};
+    for (auto pattern : patterns) {
+        SearchSignature search(arity);
+        setBits(search, pattern);
```
Comment (ohamel-softwaresecure) [inline src/test/matching_test.cpp:61]: why not just have `setBits` helper return the signature? (`auto src = setBits(arity, pat)`)

File: src/SynthesiserRelation.cpp (hunk 9 above)
Comment (SamArch27) [reply src/SynthesiserRelation.cpp:1179]: Fixed

File: src/test/matching_test.cpp (hunk 11 above)
Comment (SamArch27) [reply src/test/matching_test.cpp]: Fixed

File: src/test/matching_test.cpp (hunk 10 above)
Comment (SamArch27) [reply src/test/matching_test.cpp]: Done

File: src/SynthesiserRelation.cpp [hunk 12]
```diff
@@ -1158,8 +1161,24 @@ void SynthesiserEqrelRelation::generateTypeStruct(std::ostream& out) {
...
+                s.set(j, AttributeConstraint::Equal);
+            }
+        }
+
+        if (i & 1) {
+            s.set(0, AttributeConstraint::Equal);
+        }
+        if (i & 2) {
+            s.set(1, AttributeConstraint::Equal);
+        }
```
Comment (mmcgr) [inline src/SynthesiserRelation.cpp:1179]: Are these lines intended to still be here? It looks like they duplicate the for loop before them.

Comment (SamArch27) [reply src/SynthesiserRelation.cpp:1179]: Great catch. I'll fix this up now.

File: src/InterpreterRelation.cpp [hunk 13]
```diff
@@ -30,10 +30,8 @@ InterpreterRelation::InterpreterRelation(std::size_t arity, std::size_t auxiliar
           attributeTypes(std::move(attributeTypes)) {
     for (auto order : orderSet.getAllOrders()) {
         // Expand the order to a total order
-        std::set<int> set;
-        for (const auto& i : order) {
-            set.insert(i);
-        }
+        std::set<uint32_t> set{order.begin(), order.end()};
```
[the remark this answers is not in the document]
Comment (SamArch27) [reply src/InterpreterRelation.cpp:33]: We only use this once. Maybe if we use this in more situations, we can create a type for it?

File: src/RamIndexAnalysis.h (hunk 5 above)
Comment (ohamel-softwaresecure) [reply src/RamIndexAnalysis.h]: Super late, but might as well answer this: When you define a member func inside a class-def it's implicitly `inline`d, there's no need to explicitly say `inline` again. If you defined it in the header, but outside of the class-def, then you'd need `inline` on the definition. e.g.
```
class A { void foo() { /* okay, implicitly `inline`d b/c I'm defined inside a class-def */ } };
class B { void bar(); };
inline void B::bar() { /* explicit `inline` because I'm not defined inside a class, only declared */ }
```
